sklearn/metrics/_classification.py

```diff
@@ -2089,55 +2130,128 @@ class after being classified as negative. This is the case when the
-            warnings.warn(msg, UserWarning, stacklevel=2)
+    tn, fp, fn, tp = cm.ravel()
+    support_pos = tp + fn
+    support_neg = tn + fp
+    pos_num = tp * support_neg
+    pos_denom = fp * support_pos
+    neg_num = fn * support_neg
+    neg_denom = tn * support_pos
+
+    # if `support_pos == 0`a division by zero will occur
+    if support_pos == 0:
+        msg = (
+            "No samples of the positive class are present in `y_true`. "
+            "`positive_likelihood_ratio` and `negative_likelihood_ratio` are both set "
+            "to `np.nan`."
+        )
+        warnings.warn(msg, UndefinedMetricWarning, stacklevel=2)
         positive_likelihood_ratio = np.nan
         negative_likelihood_ratio = np.nan
-    else:
-        tn, fp, fn, tp = cm.ravel()
-        support_pos = tp + fn
-        support_neg = tn + fp
-        pos_num = tp * support_neg
-        pos_denom = fp * support_pos
-        neg_num = fn * support_neg
-        neg_denom = tn * support_pos
-
-        # If zero division warn and set scores to nan, else divide
-        if support_pos == 0:
-            msg = "no samples of the positive class were present in the testing set "
-            if raise_warning:
-                warnings.warn(msg, UserWarning, stacklevel=2)
-            positive_likelihood_ratio = np.nan
-            negative_likelihood_ratio = np.nan
-        if fp == 0:
+
+    # if `fp == 0`a division by zero will occur
+    if fp == 0:
+        if raise_warning:
             if tp == 0:
-                msg = "no samples predicted for the positive class"
+                msg_beginning = (
+                    "No samples were predicted for the positive class and "
+                    "`positive_likelihood_ratio` is "
+                )
             else:
-                msg = "positive_likelihood_ratio ill-defined and being set to nan "
-            if raise_warning:
-                warnings.warn(msg, UserWarning, stacklevel=2)
-            positive_likelihood_ratio = np.nan
+                msg_beginning = "`positive_likelihood_ratio` is ill-defined and "
+            msg_end = "set to `np.nan`. Use the `replace_undefined_by` param to "
+            "control this behavior."
+            warnings.warn(msg_beginning + msg_end, UndefinedMetricWarning, stacklevel=2)
+        if isinstance(replace_undefined_by, float) and np.isnan(replace_undefined_by):
+            positive_likelihood_ratio = replace_undefined_by
         else:
-            positive_likelihood_ratio = pos_num / pos_denom
-        if tn == 0:
-            msg = "negative_likelihood_ratio ill-defined and being set to nan "
-            if raise_warning:
-                warnings.warn(msg, UserWarning, stacklevel=2)
-            negative_likelihood_ratio = np.nan
+            # isinstance(zero_division.get("LR+", None), Real); this includes `np.inf`
+            # and `np.nan`
+            positive_likelihood_ratio = np.float64(replace_undefined_by["LR+"])
```

Comment:
```suggestion
            positive_likelihood_ratio = desired_lr_pos
```

do we need a `np.float64`  conversion? 

This also reminds me, what do we do here when we wanted to return python scalars @glemaitre

Reply:
We indeed don't need the conversion anymore (was a leftover from earlier code), and we can even use  `desired_lr_pos`. Remark: here we only get if replace_undefined_by is a dict and `desired_lr_pos` is only defined for dicts. (I have added a comment into the code so it's clear.)